### PR TITLE
Do not execute abstract tests in VintageTestEngine

### PR DESCRIPTION
**Pull request by LiamClark**
## Overview

The Junit4DiscoveryRequestResolver did not take the isPotentialJunit4TestClass
predicate in consideration. This is now added in the TestClassCollector, so
Abstract classes are properly filtered.

Multiple tests however used non-JUnit 4 compliant classes as test input.
These are now moved to make them proper JUnit 4 test classes.

Perhaps some of these test classes should be given better names, since they are now on the top level.

The logging follows the output maven generated for the same class when it was being
run from the jupiter engine.

---

I hereby agree to the terms of the JUnit Contributor License Agreement.

---

### Definition of Done

- [ ] There are no TODOs left in the code
- [ ] Method [preconditions](http://junit.org/junit5/docs/snapshot/api/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [ ] [Coding conventions](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [ ] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#tests)
- [ ] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#javadoc) and [`@API` annotations](http://junit.org/junit5/docs/snapshot/api/org/junit/platform/commons/meta/API.html)
- [ ] Change is documented in the [User Guide](http://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](http://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
- [ ] All [continuous integration builds](https://github.com/junit-team/junit5#continuous-integration-builds) pass

### CODE REVIEW

File: junit-vintage-engine/src/main/java/org/junit/vintage/engine/discovery/JUnit4DiscoveryRequestResolver.java [hunk 1]
```diff
@@ -81,7 +81,23 @@ private TestClassCollector collectTestClasses(EngineDiscoveryRequest discoveryRe
 		Filter<Class<?>> classFilter = new ExclusionReasonConsumingFilter<>(adaptedFilter,
 			(testClass, reason) -> logger.fine(() -> String.format("Class %s was excluded by a class filter: %s",
 				testClass.getName(), reason.orElse("<unknown reason>"))));
-		return collector.toRequests(classFilter.toPredicate());
+
+		Predicate<Class<?>> loggingTestClass = isPotentialJunit4Predicate();
```
Comment (sbrannen) [inline junit-vintage-engine/src/main/java/org/junit/vintage/engine/discovery/JUnit4DiscoveryRequestResolver.java]: It's not a test class that logs; rather, it's a predicate that logs the test class.

So please rename to `loggingPotentialJUnit4TestClassPredicate` or similar.

File: junit-vintage-engine/src/main/java/org/junit/vintage/engine/discovery/JUnit4DiscoveryRequestResolver.java [hunk 2]
```diff
@@ -81,7 +81,23 @@ private TestClassCollector collectTestClasses(EngineDiscoveryRequest discoveryRe
...
+
+		return collector.toRequests(classFilter.toPredicate().and(loggingTestClass));
+	}
+
+	private Predicate<Class<?>> isPotentialJunit4Predicate() {
```
Comment (sbrannen) [inline junit-vintage-engine/src/main/java/org/junit/vintage/engine/discovery/JUnit4DiscoveryRequestResolver.java]: ...is a potential JUnit 4.... what?

File: junit-vintage-engine/src/main/java/org/junit/vintage/engine/discovery/JUnit4DiscoveryRequestResolver.java [hunk 3]
```diff
@@ -81,7 +81,23 @@ private TestClassCollector collectTestClasses(EngineDiscoveryRequest discoveryRe
...
+		IsPotentialJUnit4TestClass potentialTestClass = new IsPotentialJUnit4TestClass();
```
Comment (sbrannen) [inline junit-vintage-engine/src/main/java/org/junit/vintage/engine/discovery/JUnit4DiscoveryRequestResolver.java]: No need to instantiate the predicate every time.

Please instantiate it once as a `static final` field as is done elsewhere in the code base for such predicates.

Also, please name the field consistent with the rest of the code base: `isPotentialJUnit4TestClass`

File: junit-vintage-engine/src/main/java/org/junit/vintage/engine/discovery/JUnit4DiscoveryRequestResolver.java [hunk 4]
```diff
@@ -81,7 +81,23 @@ private TestClassCollector collectTestClasses(EngineDiscoveryRequest discoveryRe
...
+		return testClass -> {
+			boolean isTestClass = potentialTestClass.test(testClass);
```
Comment (sbrannen) [inline junit-vintage-engine/src/main/java/org/junit/vintage/engine/discovery/JUnit4DiscoveryRequestResolver.java]: `isTestClass` --> `isPotentialTestClass`

File: junit-vintage-engine/src/main/java/org/junit/vintage/engine/discovery/TestClassCollector.java [hunk 5]
```diff
@@ -48,7 +48,11 @@ void addFiltered(Class<?> testClass, RunnerTestDescriptorAwareFilter filter) {
 	}
 
 	private Stream<TestClassRequest> completeRequests(Predicate<? super Class<?>> predicate) {
-		return completeTestClasses.stream().filter(predicate).map(TestClassRequest::new);
+		// @formatter:off
```
Comment (sbrannen) [inline junit-vintage-engine/src/main/java/org/junit/vintage/engine/discovery/TestClassCollector.java]: No need to reformat this.

File: junit-vintage-engine/src/test/java/org/junit/vintage/engine/VintageTestEngineExecutionTests.java [hunk 6]
```diff
@@ -66,12 +66,13 @@
...
 import org.junit.vintage.engine.samples.junit4.PlainJUnit4TestCaseWithSingleTestWhichIsIgnored;
 import org.junit.vintage.engine.samples.junit4.PlainJUnit4TestCaseWithTwoTestMethods;
+import org.junit.vintage.engine.samples.junit4.SuiteTestClass;
 import org.opentest4j.MultipleFailuresError;
 
 /**
  * @since 4.12
  */
-class VintageTestEngineExecutionTests {
+public class VintageTestEngineExecutionTests {
```
Comment (sbrannen) [inline junit-vintage-engine/src/test/java/org/junit/vintage/engine/VintageTestEngineExecutionTests.java]: Why did you make this public?

File: junit-vintage-engine/src/test/java/org/junit/vintage/engine/discovery/VintageDiscoveryRequestResolverTests.java [hunk 7]
```diff
@@ -58,19 +60,21 @@ void logsWarningWhenFilterExcludesClass() {
 			logRecord.getMessage());
 	}
 
-	public static class Foo {
```
Comment (sbrannen) [inline junit-vintage-engine/src/test/java/org/junit/vintage/engine/discovery/VintageDiscoveryRequestResolverTests.java]: Why did you extract `Foo` and `Bar` into top-level classes?

File: junit-vintage-engine/src/test/java/org/junit/vintage/engine/samples/junit4/ChildlessTestClass.java [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+
+package org.junit.vintage.engine.samples.junit4;
+
+import org.junit.runner.RunWith;
+import org.junit.vintage.engine.VintageTestEngineExecutionTests;
+
+@RunWith(VintageTestEngineExecutionTests.MisbehavingChildlessRunner.class)
+public class ChildlessTestClass {
```
Comment (sbrannen) [inline junit-vintage-engine/src/test/java/org/junit/vintage/engine/samples/junit4/ChildlessTestClass.java]: Why did you make this a top-level class instead of simply a nested static class in the test class that actually uses it?

File: junit-vintage-engine/src/test/java/org/junit/vintage/engine/samples/junit4/SuiteTestClass.java [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+ *
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+
+package org.junit.vintage.engine.samples.junit4;
+
+import org.junit.runner.RunWith;
+import org.junit.vintage.engine.VintageTestEngineExecutionTests;
+
+@RunWith(VintageTestEngineExecutionTests.MisbehavingSuiteRunner.class)
```
Comment (sbrannen) [inline junit-vintage-engine/src/test/java/org/junit/vintage/engine/samples/junit4/SuiteTestClass.java]: Why did you make this a top-level class instead of simply a nested static class in the test class that actually uses it?

**Comment by sbrannen** [comment]
Commit message must follow the guidelines laid out here: https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#commit-messages

In addition, the commit message must contain a reference to the issue it addresses -- for example:

```
Issue: #999
```

File: junit-vintage-engine/src/test/java/org/junit/vintage/engine/discovery/VintageDiscoveryRequestResolverTests.java (hunk 7 above)
Comment (LiamClark) [inline junit-vintage-engine/src/test/java/org/junit/vintage/engine/discovery/VintageDiscoveryRequestResolverTests.java]: @sbrannen Otherwise they would not be discovered by the vintage engine. 
The isPotentialJunit4TestClass predicate contains a check: !candidate.isMemberClass();
So foo as a nested class would be filtered out making the tests fail.

**Comment by sbrannen** [comment]
Although the title "Fix Junit4DiscoveryRequestResolver" is technically correct, it would be preferable to describe what is being fixed in plain English.

Something like the following would be better: _Do not execute abstract tests in VintageTestEngine_

**Comment by sbrannen** [comment]
Hi @LiamClark,

Thanks for the PR!

I've left some comments and performed a quick review.

File: junit-vintage-engine/src/main/java/org/junit/vintage/engine/discovery/TestClassCollector.java (hunk 5 above)
Comment (LiamClark) [inline junit-vintage-engine/src/main/java/org/junit/vintage/engine/discovery/TestClassCollector.java]: Ah yes my bad, I added an extra filter here in an initial try, but forgot to remove the formatting

File: junit-vintage-engine/src/main/java/org/junit/vintage/engine/discovery/JUnit4DiscoveryRequestResolver.java (hunk 3 above)
Comment (jbduncan) [inline junit-vintage-engine/src/main/java/org/junit/vintage/engine/discovery/JUnit4DiscoveryRequestResolver.java]: Consider making the predicate class an enum like the following, since then it cannot be instantiated more than once.
```java
public enum IsPotentialJUnit4TestClass {
  INSTANCE;
  . . .
}
```

Comment (LiamClark) [inline junit-vintage-engine/src/main/java/org/junit/vintage/engine/discovery/JUnit4DiscoveryRequestResolver.java]: @jbduncan @sbrannen doesn't it need to be instantiated for every instance of JUnit4DiscoveryRequestResolver since it accesses the logger field?

Comment (jbduncan) [inline junit-vintage-engine/src/main/java/org/junit/vintage/engine/discovery/JUnit4DiscoveryRequestResolver.java]: Actually, this _may_ be a bad idea if other predicate classes exist that follow the same convention as `IsPotentialJUnit4TestClass` currently does. (I can't quite check ATM since I'm a bit busy...)

File: junit-vintage-engine/src/test/java/org/junit/vintage/engine/discovery/VintageDiscoveryRequestResolverTests.java (hunk 7 above)
Comment (sbrannen) [inline junit-vintage-engine/src/test/java/org/junit/vintage/engine/discovery/VintageDiscoveryRequestResolverTests.java]: No. That should not be the case. A static nested class is perfectly valid as a test class in the eyes of JUnit 4. And a "member class" is non-static. 

But... if you believe otherwise, please submit a separate PR demonstrating the bug you think is lurking here.

In the interim, please revert your changes to the `Foo` and `Bar` classes.

File: junit-vintage-engine/src/main/java/org/junit/vintage/engine/discovery/JUnit4DiscoveryRequestResolver.java (hunk 3 above)
Comment (sbrannen) [inline junit-vintage-engine/src/main/java/org/junit/vintage/engine/discovery/JUnit4DiscoveryRequestResolver.java]: No, `IsPotentialJUnit4TestClass` does _not_ access any logger whatsoever.

Comment (sbrannen) [inline junit-vintage-engine/src/main/java/org/junit/vintage/engine/discovery/JUnit4DiscoveryRequestResolver.java]: Please do not make the predicate an enum. We don't do that anywhere, and if we were to start implementing our predicates as _singletons_, we would do that in a separate issue.

File: junit-vintage-engine/src/test/java/org/junit/vintage/engine/discovery/VintageDiscoveryRequestResolverTests.java (hunk 7 above)
Comment (marcphilipp) [inline junit-vintage-engine/src/test/java/org/junit/vintage/engine/discovery/VintageDiscoveryRequestResolverTests.java]: Since #787 has now been merged this PR should be rebased onto master and this particular change can be reverted.

**Comment by LiamClark** [comment]
@sbrannen I've rebased and undid the move for the test classes to the top level.
The tests also used two local classes. These are now public static member classes.

is this ok?

**Comment by sbrannen** [comment]
Yep, looks good now.

Thanks! 👍
